### PR TITLE
[deploy]: AWS Elastic

### PR DESCRIPTION
**Pull request by Izak88**

### CODE REVIEW

File: docs/DEPLOY_ELASTIC.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+Abstruse CI can automatically deploy to Amazon Elastic Beanstalk after a successful build.
+
+For a minimal configuration, you can add the following to your config file `.abstruse.yml`.
+
+```
+deploy:
+  - provider: elastic
+  - application: "test-application"
+  - accessKeyId: "your_aws_access_key"
```
Comment (jkuri) [inline docs/DEPLOY_ELASTIC.md]: you provide key with the configuration? I guess everyone can see it then.

File: src/api/deploy/aws-elastic.ts [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+import { findFromEnvVariables } from '../deploy';
+import * as style from 'ansi-styles';
+import chalk from 'chalk';
+
+export function elasticDeploy(
+  preferences: any, container: string, variables: string[]
+): Observable<any> {
+  return new Observable((observer: Observer<any>) => {
+
+    console.log(JSON.stringify(preferences));
```
Comment (jkuri) [inline src/api/deploy/aws-elastic.ts]: console.log

File: docs/DEPLOY_ELASTIC.md (hunk 1 above)
Comment (Izak88) [reply docs/DEPLOY_ELASTIC.md]: That's just an example, read the next section, where it says that it's not recommended and that the best way is through environment variables...

File: src/api/deploy/aws-elastic.ts [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+      if (!applicationType) {
+        applicationType = 'zip';
+      }
+
+      let msg = style.yellow.open + style.bold.open + '==> deploy started' +
+      style.bold.close + style.yellow.close + '\r\n';
+      observer.next({ type: 'data', data: msg });
+
+      // 2. install awscli and set credentials
+      let command = { type: CommandType.deploy, command: 'sudo apt-get install awscli -y' };
```
Comment (jkuri) [inline src/api/deploy/aws-elastic.ts]: is it a good thing to hardcode `sudo apt-get`? `apt-get` is included in debian-like Linux OS-es, but what if you use alpine for example?

File: src/api/deploy/aws-elastic.ts [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+      attachExec(container, command)
+        .toPromise()
+        .then(result => {
+          if (!(result && result.data === 0)) {
+            const msg = `apt-get install awscli failed`;
+            observer.next({ type: 'containerError', data: msg});
+            return Promise.reject(-1);
```
Comment (jkuri) [inline src/api/deploy/aws-elastic.ts]: what does `-1` means?

File: src/api/deploy/aws-elastic.ts [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+                  + ` --solution-stack-name "${solutionStackName}"`
+              };
+
+              return attachExec(container, command)
+                .toPromise()
+                .then(result => {
+                  if (!(result && result.data === 0)) {
+                    const msg = 'aws create environment failed';
+                    observer.next({ type: 'containerError', data: msg});
+                    return Promise.reject(-1);
```
Comment (jkuri) [inline src/api/deploy/aws-elastic.ts]: here too, if you are trying to produce an error that would be anything but 0.
0 means success and 1-n means error.

Comment (Izak88) [reply src/api/deploy/aws-elastic.ts]: I know, what would you suggest instead -1? Like you said, I just want to produce error.

Comment (jkuri) [reply src/api/deploy/aws-elastic.ts]: you can do `new Error('description')`

**Comment by Izak88** [comment]
Updated PR after code review...

File: src/api/deploy/aws-elastic.ts [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+            + style.bold.close + style.yellow.close + '\r\n';
+          observer.next({ type: 'data', data: msg });
+          observer.complete();
+        })
+        .catch(err => {
+          observer.error(err);
+          observer.complete();
+        });
+    } else {
+      observer.error(-1);
```
Comment (jkuri) [inline src/api/deploy/aws-elastic.ts]: still having this `-1` that we talked about.

Comment (Izak88) [reply src/api/deploy/aws-elastic.ts]: done

**Pull request merged**